### PR TITLE
HHH-18446 Cannot use @JdbcTypeCode(SqlTypes.LONG32VARBINARY) on field with null value

### PR DESCRIPTION
**Pull request by cigaly**
Jira issue [HHH-18446](https://hibernate.atlassian.net/browse/HHH-18446)

Fixing problem by adding (configurable) default SQL type code to LongVarbinaryJdbcType abd LongVarcharJdbcType
----------------------
By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt)
and can be relicensed under the terms of the [LGPL v2.1 license](https://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt) in the future at the maintainers' discretion.
For more information on licensing, please check [here](https://github.com/hibernate/hibernate-orm/blob/main/CONTRIBUTING.md#legal).

----------------------

[HHH-18446]: https://hibernate.atlassian.net/browse/HHH-18446?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/type/descriptor/jdbc/internal/JdbcTypeBaseline.java [hunk 1]
```diff
@@ -85,12 +85,12 @@ public static void prime(BaselineTarget target) {
 		target.addDescriptor( BinaryJdbcType.INSTANCE );
 		target.addDescriptor( VarbinaryJdbcType.INSTANCE );
 		target.addDescriptor( LongVarbinaryJdbcType.INSTANCE );
-		target.addDescriptor( new LongVarbinaryJdbcType(SqlTypes.LONG32VARBINARY) );
+		target.addDescriptor( new LongVarbinaryJdbcType(SqlTypes.LONGVARBINARY, SqlTypes.LONG32VARBINARY) );
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/type/descriptor/jdbc/internal/JdbcTypeBaseline.java:88]: @cigaly this looks fine, but would a simpler (and maybe slightly better) change be to just register the type twice, something like:

```java
target.addDescriptor( new LongVarbinaryJdbcType(SqlTypes.SqlTypes.LONG32VARBINARY) );
target.addDescriptor( SqlTypes.LONG32VARBINARY, new LongVarbinaryJdbcType(SqlTypes.LONG32VARBINARY) );
```

(I'm not sure, just asking.)

Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/type/descriptor/jdbc/internal/JdbcTypeBaseline.java:88]: Oh, no, wait, scratch that, I misunderstood.

File: hibernate-core/src/main/java/org/hibernate/type/descriptor/jdbc/LongVarcharJdbcType.java [hunk 2]
```diff
@@ -39,6 +46,11 @@ public int getJdbcTypeCode() {
 		return jdbcTypeCode;
 	}
 
+	@Override
+	public int getDefaultSqlTypeCode() {
+		return defaultSqlTypeCode;
+	}
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/type/descriptor/jdbc/LongVarcharJdbcType.java:52]: Now that I understand, I would say we don't need to parameterize the class by `defaultSqlTypeCode`, we can just hardcode this method to return `Types.LONGVARCHAR`, right?

Comment (cigaly) [reply hibernate-core/src/main/java/org/hibernate/type/descriptor/jdbc/LongVarcharJdbcType.java]: I guess that this is, at least, partially correct. Only it seems that JDBC type code should be fixed, and default SQL code changeable, but I will check that again.

Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/type/descriptor/jdbc/LongVarcharJdbcType.java]: The one that is used for binding to `null` values to JDBC is that one that needs to be hardcoded. The one used for registration and for schema export is the one that's parameterized.

Comment (cigaly) [reply hibernate-core/src/main/java/org/hibernate/type/descriptor/jdbc/LongVarcharJdbcType.java]: I guess now that is exactly this

File: hibernate-core/src/main/java/org/hibernate/type/descriptor/jdbc/LongVarbinaryJdbcType.java [hunk 3]
```diff
@@ -14,14 +14,19 @@
...
+	private final int defaultSqlTypeCode;
 
 	public LongVarbinaryJdbcType() {
-		this(Types.LONGVARBINARY);
+		this( Types.LONGVARBINARY );
 	}
 
-	public LongVarbinaryJdbcType(int jdbcTypeCode) {
-		this.jdbcTypeCode = jdbcTypeCode;
+	public LongVarbinaryJdbcType(final int defaultSqlTypeCode) {
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/type/descriptor/jdbc/LongVarbinaryJdbcType.java:23]:
```suggestion
	public LongVarbinaryJdbcType(int defaultSqlTypeCode) {
```

We don't usually bother sticking `final` on parameters, even though it makes sense.

**Review by gavinking — APPROVED** [review]
This looks good now.

**Review by gavinking — APPROVED** [review]

**Comment by gavinking** [comment]
@cigaly why is this still in Draft mode?

**Comment by cigaly** [comment]
Not any more :-)

**Comment by cigaly** [comment]
Now I've switched back from draft mode

**Comment by gavinking** [comment]
Actually, @cigaly can you squash that last commit please?

**Comment by gavinking** [comment]
Thank you so much! 🙏🙏

**Pull request merged**